### PR TITLE
Added WavefrontHistogram and unit tests

### PR DESCRIPTION
**Pull request by hanwavefront**

### CODE REVIEW

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogram.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+   * @return returns aggregated collection of all the bins prior to the current minute
+   */
+  public List<MinuteBin> bins(boolean clear) {
+    final List<MinuteBin> result = new ArrayList<>();
+    final long cutoffMillis = currentMinuteMillis();
+
+    perThreadHistogramBins.asMap().values().stream().flatMap(List::stream).
+        filter(i -> i.getMinuteMillis() < cutoffMillis).forEach(result::add);
+
+    if (clear) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogram.java]: let clear be true by default so no need of if condition

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogram.java [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+  }
+
+  /**
+   * Aggregates all the bins prior to the current minute
+   * This is because threads might be updating the current minute bin while the bins() method is invoked
+   *
+   * @param clear if set to true, will clear all values in bins prior to the current minute
+   * @return returns aggregated collection of all the bins prior to the current minute
+   */
+  public List<MinuteBin> bins(boolean clear) {
```
Comment (sushantdewan123) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogram.java]: I don't think we should give the boolean option "clear" to the users of this SDK.
Any aggregated values prior to the current min should be cleared by default.

Also, to save some work for the Tier2 SDK, we should rename this method to flushDistributions() and should return a list of distribution strings - List<Pair<Long, String> centroids
where the 2 tuple Pair represents - distribution

that way the Tier 2 client does not have to do this and need not know what MinuteBin is -

for (MinuteBin minuteBin : bins) {
          List<Pair<Double, Integer>> centroids = new ArrayList<>();
          for (Centroid c : minuteBin.getDist().centroids()) {
            centroids.add(new Pair<>(c.mean(), c.count()));
          }
          wavefront.send(histogramGranularities, minuteBin.getMinuteMillis() / 1000,
              centroids, prefixAndSanitize(name), source, pointTags);
        }

Also mention in the javadoc that flushDistributions() has side effects (i.e. it is not read-only method) and will mutate the state of your system.

**Comment by sushantdewan123** [comment]
@panghy 

We are adding 3 more dependencies in this PR
```
        <dependency>
            <groupId>com.tdunning</groupId>
            <artifactId>t-digest</artifactId>
            <version>3.2</version>
        </dependency>
        <dependency>
            <groupId>com.github.ben-manes.caffeine</groupId>
            <artifactId>caffeine</artifactId>
            <version>2.6.2</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>org.hamcrest</groupId>
            <artifactId>java-hamcrest</artifactId>
            <version>RELEASE</version>
            <scope>test</scope>
        </dependency>
```
Let me know if we need to shade/repackage any of those.

**Comment by sushantdewan123** [comment]
Renamed to "WavefrontHistogramImpl" to avoid conflict Tier 2 com.codahale.metrics.WavefrontHistogram that uses this implementation as a delegate.

**Comment by panghy** [comment]
Please use ThreadLocal for List of MinuteBins and use a collection of weakreferences that you can check on reporting (and purge).

File: src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+        currentBin.distribution.add(means.get(i), counts.get(i));
+      }
+    }
+  }
+
+  /**
+   * @return returns the number of values in the distribution.
+   */
+  public long getCount() {
+    return globalHistogramBinsList.stream().map(Reference::get).filter(Objects::nonNull).
```
Comment (panghy) [inline src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java]: access to globalHistogramBinList is not protected from multiple threads accessing it. Can probably use a Set that's backed by a Sets.fromMap(new ConcurrentHashMap...), have synchronization or CopyOnWriteArrayList

**Pull request merged**